File: src/game/piece/piece.ts
import { colors } from '~/game/consts/board';

// Interface for move behavior
export interface MoveBehavior {
  initializeAttacks(): void;
}

// Leaper move behavior
export class LeaperMoveBehavior implements MoveBehavior {
  private leaperOffsets: number[][][];
  private leaperPieceStateLo: Uint32Array[][];
  private leaperPieceStateHi: Uint32Array[][];

  constructor(leaperOffsets: number[][][]) {
    this.leaperOffsets = leaperOffsets;
    this.leaperPieceStateLo = Array.from({ length: 2 }, () =>
      Array.from({ length: leaperOffsets.length }, () => new Uint32Array(64)),
    );
    this.leaperPieceStateHi = Array.from({ length: 2 }, () =>
      Array.from({ length: leaperOffsets.length }, () => new Uint32Array(64)),
    );
  }

  /**
   * Initialize leaper piece attacks.
   */
  initializeAttacks(): void {
    for (let i = 0; i < this.leaperOffsets.length; i++) {
      for (let square = 0; square < 64; square++) {
        const white = this.maskLeaperAttacks(square, this.leaperOffsets[i], colors.WHITE);
        const black = this.maskLeaperAttacks(square, this.leaperOffsets[i], colors.BLACK);
        this.leaperPieceStateLo[colors.WHITE][i][square] = white.lo;
        this.leaperPieceStateHi[colors.WHITE][i][square] = white.hi;
        this.leaperPieceStateLo[colors.BLACK][i][square] = black.lo;
        this.leaperPieceStateHi[colors.BLACK][i][square] = black.hi;
      }
    }
  }

  /**
   * Low 32 bits of the leaper attack table.
   */
  getLeaperPieceStateLo(): Uint32Array[][] {
    return this.leaperPieceStateLo;
  }

  /**
   * High 32 bits of the leaper attack table.
   */
  getLeaperPieceStateHi(): Uint32Array[][] {
    return this.leaperPieceStateHi;
  }

  /**
   * Mask leaper attacks.
   */
  private maskLeaperAttacks(
    pos: number,
    offsets: number[][],
    color: number,
  ): { lo: number; hi: number } {
    let lo = 0;
    let hi = 0;
    const rankTop = pos >> 3;
    const file = pos & 7;
    const sign = color === colors.BLACK ? -1 : 1;
    for (let o = 0; o < offsets.length; o++) {
      const df = offsets[o][0];
      const dr = offsets[o][1];
      const nf = file + sign * df;
      const nrt = rankTop - sign * dr;
      if (nf < 0 || nf > 7 || nrt < 0 || nrt > 7) continue;
      const sq = nrt * 8 + nf;
      if (sq < 32) lo |= 1 << sq;
      else hi |= 1 << (sq - 32);
    }
    return { lo, hi };
  }

  // Getters
  getLeaperOffsets(): number[][][] {
    return this.leaperOffsets;
  }

  // Setters
  setLeaperOffsets(offsets: number[][][]): void {
    this.leaperOffsets = offsets;
  }
}

// Pawn move behavior
export class PawnMoveBehavior implements MoveBehavior {
  private pawnPieceStateLo: Uint32Array[];
  private pawnPieceStateHi: Uint32Array[];

  constructor() {
    this.pawnPieceStateLo = Array.from({ length: 2 }, () => new Uint32Array(64));
    this.pawnPieceStateHi = Array.from({ length: 2 }, () => new Uint32Array(64));
  }

  /**
   * Initialize pawn piece attacks.
   */
  initializeAttacks(): void {
    for (let square = 0; square < 64; square++) {
      const white = this.maskPawnAttacks(colors.WHITE, square);
      const black = this.maskPawnAttacks(colors.BLACK, square);
      this.pawnPieceStateLo[colors.WHITE][square] = white.lo;
      this.pawnPieceStateHi[colors.WHITE][square] = white.hi;
      this.pawnPieceStateLo[colors.BLACK][square] = black.lo;
      this.pawnPieceStateHi[colors.BLACK][square] = black.hi;
    }
  }

  /**
   * Low 32 bits of the pawn attack table.
   */
  getPawnPieceStateLo(): Uint32Array[] {
    return this.pawnPieceStateLo;
  }

  /**
   * High 32 bits of the pawn attack table.
   */
  getPawnPieceStateHi(): Uint32Array[] {
    return this.pawnPieceStateHi;
  }

  /**
   * Mask pawn attacks.
   */
  private maskPawnAttacks(color: number, pos: number): { lo: number; hi: number } {
    let lo = 0;
    let hi = 0;
    const rankTop = pos >> 3;
    const file = pos & 7;
    // white captures toward rankTop - 1, black toward rankTop + 1
    const nrt = color === colors.WHITE ? rankTop - 1 : rankTop + 1;
    if (nrt < 0 || nrt > 7) return { lo, hi };
    if (file - 1 >= 0) {
      const sq = nrt * 8 + (file - 1);
      if (sq < 32) lo |= 1 << sq;
      else hi |= 1 << (sq - 32);
    }
    if (file + 1 <= 7) {
      const sq = nrt * 8 + (file + 1);
      if (sq < 32) lo |= 1 << sq;
      else hi |= 1 << (sq - 32);
    }
    return { lo, hi };
  }
}

/**
 * Pack the occupancy bits at the given positions into a compact index.
 */
function pextIndex(occLo: number, occHi: number, bits: number[]): number {
  let idx = 0;
  for (let k = 0; k < bits.length; k++) {
    const sq = bits[k];
    if (sq < 32) {
      if ((occLo >>> sq) & 1) idx |= 1 << k;
    } else {
      if ((occHi >>> (sq - 32)) & 1) idx |= 1 << k;
    }
  }
  return idx;
}

// Sliding move behavior
export class SlidingMoveBehavior implements MoveBehavior {
  private straight: boolean;
  private diagonal: boolean;
  // attack tables (lo/hi per square) and the relevant mask bit positions
  private straightAttackLo: Uint32Array[];
  private straightAttackHi: Uint32Array[];
  private straightMaskBits: number[][];
  private diagonalAttackLo: Uint32Array[];
  private diagonalAttackHi: Uint32Array[];
  private diagonalMaskBits: number[][];

  constructor(straight: boolean, diagonal: boolean) {
    this.straight = straight;
    this.diagonal = diagonal;
    this.straightAttackLo = [];
    this.straightAttackHi = [];
    this.straightMaskBits = [];
    this.diagonalAttackLo = [];
    this.diagonalAttackHi = [];
    this.diagonalMaskBits = [];
  }

  /**
   * Initialize sliding piece attacks.
   */
  initializeAttacks(): void {
    // sliding pieces initialize their attack tables externally
  }

  /**
   * Get sliding piece attacks as {lo, hi}.
   */
  getSliderAttacksLoHi(pos: number, occLo: number, occHi: number): { lo: number; hi: number } {
    let lo = 0;
    let hi = 0;
    if (this.straight) {
      const idx = pextIndex(occLo, occHi, this.straightMaskBits[pos]);
      lo |= this.straightAttackLo[pos][idx];
      hi |= this.straightAttackHi[pos][idx];
    }
    if (this.diagonal) {
      const idx = pextIndex(occLo, occHi, this.diagonalMaskBits[pos]);
      lo |= this.diagonalAttackLo[pos][idx];
      hi |= this.diagonalAttackHi[pos][idx];
    }
    return { lo, hi };
  }

  // Getters
  getStraight(): boolean {
    return this.straight;
  }

  getDiagonal(): boolean {
    return this.diagonal;
  }

  // Setters
  setStraight(straight: boolean): void {
    this.straight = straight;
  }

  setDiagonal(diagonal: boolean): void {
    this.diagonal = diagonal;
  }

  setStraightAttackTables(lo: Uint32Array[], hi: Uint32Array[], bits: number[][]): void {
    this.straightAttackLo = lo;
    this.straightAttackHi = hi;
    this.straightMaskBits = bits;
  }

  setDiagonalAttackTables(lo: Uint32Array[], hi: Uint32Array[], bits: number[][]): void {
    this.diagonalAttackLo = lo;
    this.diagonalAttackHi = hi;
    this.diagonalMaskBits = bits;
  }
}

// Piece base class
abstract class Piece {
  private id: number;
  private color: number;
  private move: number;
  private globalEffect: boolean;
  private firstMove: number;
  private king: boolean;
  private pawn: boolean;
  private enpassant: boolean;
  private promote: boolean;
  private slider: boolean;
  private leaper: boolean;
  private moveBehavior: MoveBehavior;
  private pieceMask: BigUint64Array;
  private rotationalMoveType: string;
  private reverse: Map<number, number>;

  constructor(
    id: number,
    color: number,
    moveBehavior: MoveBehavior,
    rotationalMoveType: string = 'DEFAULT',
  ) {
    this.id = id;
    this.color = color;
    this.move = 0;
    this.globalEffect = false;
    this.firstMove = -1;
    this.king = false;
    this.pawn = false;
    this.enpassant = false;
    this.promote = false;
    this.slider = false;
    this.leaper = false;
    this.moveBehavior = moveBehavior;
    this.pieceMask = new BigUint64Array(64);
    this.rotationalMoveType = rotationalMoveType;
    this.reverse = new Map<number, number>();
    this.moveBehavior.initializeAttacks();
  }

  // Getters
  getId(): number {
    return this.id;
  }

  getColor(): number {
    return this.color;
  }

  getMove(): number {
    return this.move;
  }

  getGlobalEffect(): boolean {
    return this.globalEffect;
  }

  getFirstMove(): number {
    return this.firstMove;
  }

  getKing(): boolean {
    return this.king;
  }

  getPawn(): boolean {
    return this.pawn;
  }

  getEnpassant(): boolean {
    return this.enpassant;
  }

  getPromote(): boolean {
    return this.promote;
  }

  getSlider(): boolean {
    return this.slider;
  }

  getLeaper(): boolean {
    return this.leaper;
  }

  getMoveBehavior(): MoveBehavior {
    return this.moveBehavior;
  }

  getPieceMask(): BigUint64Array {
    return this.pieceMask;
  }

  getRotationalMoveType(): string {
    return this.rotationalMoveType;
  }

  getReverse(): Map<number, number> {
    return this.reverse;
  }

  // Setters
  setId(id: number): void {
    this.id = id;
  }

  setColor(color: number): void {
    this.color = color;
  }

  setMove(move: number): void {
    this.move = move;
  }

  setGlobalEffect(effect: boolean): void {
    this.globalEffect = effect;
  }

  setFirstMove(move: number): void {
    this.firstMove = move;
  }

  setKing(king: boolean): void {
    this.king = king;
  }

  setPawn(pawn: boolean): void {
    this.pawn = pawn;
  }

  setEnpassant(enpassant: boolean): void {
    this.enpassant = enpassant;
  }

  setPromote(promote: boolean): void {
    this.promote = promote;
  }

  setSlider(slider: boolean): void {
    this.slider = slider;
  }

  setLeaper(leaper: boolean): void {
    this.leaper = leaper;
  }

  setMoveBehavior(behavior: MoveBehavior): void {
    this.moveBehavior = behavior;
  }

  setPieceMask(mask: BigUint64Array): void {
    this.pieceMask = mask;
  }

  setRotationalMoveType(type: string): void {
    this.rotationalMoveType = type;
  }

  setReverse(reverse: Map<number, number>): void {
    this.reverse = reverse;
  }
}

export class KingPiece extends Piece {
  constructor(id: number, color: number) {
    super(
      id,
      color,
      new LeaperMoveBehavior([
        [
          [-1, 1],
          [0, 1],
          [1, 1],
          [-1, 0],
          [1, 0],
          [-1, -1],
          [0, -1],
          [1, -1],
        ],
      ]),
    );
    this.setKing(true);
    this.setLeaper(true);
  }
}

export class QueenPiece extends Piece {
  constructor(id: number, color: number) {
    super(id, color, new SlidingMoveBehavior(true, true));
    this.setSlider(true);
  }
}

export class RookPiece extends Piece {
  constructor(id: number, color: number) {
    super(id, color, new SlidingMoveBehavior(true, false));
    this.setSlider(true);
  }
}

export class BishopPiece extends Piece {
  constructor(id: number, color: number) {
    super(id, color, new SlidingMoveBehavior(false, true));
    this.setSlider(true);
  }
}

export class KnightPiece extends Piece {
  constructor(id: number, color: number) {
    super(
      id,
      color,
      new LeaperMoveBehavior([
        [
          [-2, 1],
          [-1, 2],
          [1, 2],
          [2, 1],
          [-2, -1],
          [-1, -2],
          [1, -2],
          [2, -1],
        ],
      ]),
    );
    this.setLeaper(true);
  }
}

export class PawnPiece extends Piece {
  constructor(id: number, color: number) {
    super(id, color, new PawnMoveBehavior());
    this.setPawn(true);
    this.setEnpassant(true);
    this.setPromote(true);
  }
}

export class BalloonPiece extends Piece {
  constructor(id: number, color: number) {
    super(
      id,
      color,
      new LeaperMoveBehavior([
        [
          [0, 2],
          [0, -2],
          [1, 1],
          [1, -1],
          [-1, 1],
          [-1, -1],
        ],
        [
          [0, 2],
          [0, -2],
          [1, 2],
          [1, -2],
          [2, 1],
          [2, -1],
          [-1, 2],
          [-1, -2],
          [-2, 1],
          [-2, -1],
        ],
      ]),
      'ROTATE',
    );
    this.setLeaper(true);
  }
}

export class SpongebobPiece extends Piece {
  constructor(id: number, color: number) {
    super(
      id,
      color,
      new LeaperMoveBehavior([
        [
          [1, 0],
          [-1, 0],
          [0, 1],
          [0, -1],
          [1, 1],
          [1, -1],
          [-1, 1],
          [-1, -1],
        ],
        [
          [2, 0],
          [-2, 0],
          [0, 2],
          [0, -2],
          [2, 2],
          [2, -2],
          [-2, 2],
          [-2, -2],
        ],
        [
          [3, 0],
          [-3, 0],
          [0, 3],
          [0, -3],
          [3, 3],
          [3, -3],
          [-3, 3],
          [-3, -3],
        ],
        [
          [2, 0],
          [-2, 0],
          [0, 2],
          [0, -2],
          [2, 2],
          [2, -2],
          [-2, 2],
          [-2, -2],
        ],
      ]),
      'ROTATE',
    );
    this.setLeaper(true);
  }
}

export class PogoPiece extends Piece {
  constructor(id: number, color: number) {
    super(
      id,
      color,
      new LeaperMoveBehavior([[[0, 2]], [[0, -1]], [[0, 1]], [[0, -2]]]),
      'REVERSE_ROTATE',
    );
    this.setLeaper(true);
  }
}

// Piece factory
class PieceFactory {
  static createPiece(type: string, id: number, color: number): Piece {
    switch (type.toLowerCase()) {
      case 'king':
        return new KingPiece(id, color);
      case 'queen':
        return new QueenPiece(id, color);
      case 'rook':
        return new RookPiece(id, color);
      case 'bishop':
        return new BishopPiece(id, color);
      case 'knight':
        return new KnightPiece(id, color);
      case 'pawn':
        return new PawnPiece(id, color);
      case 'balloon':
        return new BalloonPiece(id, color);
      case 'pogo':
        return new PogoPiece(id, color);
      case 'spongebob':
        return new SpongebobPiece(id, color);
      default:
        throw new Error(`Unknown piece type: ${type}`);
    }
  }
}

export { Piece, PieceFactory };

File: server/piece.cpp
#include <stdint.h>
#include <cstddef>

#include "generator.hpp"
#include "piece.hpp"

#define NOT_A_FILE 18374403900871474942ULL
#define NOT_AB_FILE 18229723555195321596ULL
#define NOT_H_FILE 9187201950435737471ULL
#define NOT_HG_FILE 4557430888798830399ULL

namespace piece {
  Piece::Piece(int id, int color) {
    this->id = id;
    this->color = color;

    this->move = 0, this->king = 0, this->pawn = 0,
    this->enpassant = 0, this->promote = 0;

    this->slider = 0, this->leaper = 0;        
    this->straight = 0, this->diagonal = 0;

    this->rotational_move_type = 0;
    this->leaper_offsets = LeaperOffsets();
    this->straight_constraints = nullptr;
    this->diagonal_constraints = nullptr;

    this->piece_mask = nullptr;
    this->straight_piece_mask = nullptr;
    this->diagonal_piece_mask = nullptr;

    this->sliding_diagonal_piece_state = nullptr;
    this->sliding_straight_piece_state = nullptr;
    this->leaper_piece_state = nullptr;
    this->pawn_piece_state = nullptr;
  }

  Piece::~Piece() {
    delete[] this->straight_constraints;
    delete[] this->diagonal_constraints;
    delete[] this->piece_mask;
    delete[] this->straight_piece_mask;
    delete[] this->diagonal_piece_mask;
    delete[] this->sliding_diagonal_piece_state;
    delete[] this->sliding_straight_piece_state;
    delete[] this->leaper_piece_state;
    delete[] this->pawn_piece_state;
  }

  Pawn::Pawn(int id, int color) : Piece(id, color) {
    this->pawn = 1;
    this->enpassant = 1;
    this->promote = 1;
  }

  Knight::Knight(int id, int color) : Piece(id, color) {
    this->leaper = 1;
    this->leaper_offsets.offsets = new int**[1];

    this->leaper_offsets.offsets[0] = new int*[8];
    for (int i = 0; i < 8; ++i) {
      this->leaper_offsets.offsets[0][i] = new int[2];
    }

    /* KNIGHT MOVES */
    this->leaper_offsets.offsets[0][0][0] = -2; this->leaper_offsets.offsets[0][0][1] = 1;
    this->leaper_offsets.offsets[0][1][0] = -1; this->leaper_offsets.offsets[0][1][1] = 2;
    this->leaper_offsets.offsets[0][2][0] = 1; this->leaper_offsets.offsets[0][2][1] = 2;
    this->leaper_offsets.offsets[0][3][0] = 2; this->leaper_offsets.offsets[0][3][1] = 1;
    this->leaper_offsets.offsets[0][4][0] = -2; this->leaper_offsets.offsets[0][4][1] = -1;
    this->leaper_offsets.offsets[0][5][0] = -1; this->leaper_offsets.offsets[0][5][1] = -2;
    this->leaper_offsets.offsets[0][6][0] = 1; this->leaper_offsets.offsets[0][6][1] = -2;
    this->leaper_offsets.offsets[0][7][0] = 2; this->leaper_offsets.offsets[0][7][1] = -1;

    this->leaper_offsets.size = 1;
  }

  Knight::~Knight() {
    for (int i = 0; i < 8; ++i) {
      delete[] this->leaper_offsets.offsets[0][i];
    }
    delete[] this->leaper_offsets.offsets[0];
  }

  Bishop::Bishop(int id, int color) : Piece(id, color) {
    this->slider = 1;
    this->diagonal = 1;
    this->diagonal_constraints = new int[4]{8, 8, 8, 8};
  }

  Bishop::~Bishop() {
    delete[] this->diagonal_constraints;
  }

  Rook::Rook(int id, int color) : Piece(id, color) {
    this->slider = 1;
    this->straight = 1;
    this->straight_constraints = new int[4]{8, 8, 8, 8};
  }

  Rook::~Rook() {
    delete[] this->straight_constraints;
  }

  Queen::Queen(int id, int color) : Piece(id, color) {
    this->slider = 1;
    this->straight = 1;
    this->diagonal = 1;

    this->straight_constraints = new int[4]{8, 8, 8, 8};
    this->diagonal_constraints = new int[4]{8, 8, 8, 8};
  }

  Queen::~Queen() {
    delete[] this->straight_constraints;
    delete[] this->diagonal_constraints;
  }

  King::King(int id, int color) : Piece(id, color) {
    this->leaper = 1;
    this->leaper_offsets.offsets = new int**[1];

    this->leaper_offsets.offsets[0] = new int*[8];
    for (int i = 0; i < 8; ++i) {
      this->leaper_offsets.offsets[0][i] = new int[2];
    }

    /* KING MOVES */
    this->leaper_offsets.offsets[0][0][0] = -1; this->leaper_offsets.offsets[0][0][1] = 1;
    this->leaper_offsets.offsets[0][1][0] = 0; this->leaper_offsets.offsets[0][1][1] = 1;
    this->leaper_offsets.offsets[0][2][0] = 1; this->leaper_offsets.offsets[0][2][1] = 1;
    this->leaper_offsets.offsets[0][3][0] = -1; this->leaper_offsets.offsets[0][3][1] = 0;
    this->leaper_offsets.offsets[0][4][0] = 1; this->leaper_offsets.offsets[0][4][1] = 0;
    this->leaper_offsets.offsets[0][5][0] = -1; this->leaper_offsets.offsets[0][5][1] = -1;
    this->leaper_offsets.offsets[0][6][0] = 0; this->leaper_offsets.offsets[0][6][1] = -1;
    this->leaper_offsets.offsets[0][7][0] = 1; this->leaper_offsets.offsets[0][7][1] = -1;

    this->leaper_offsets.size = 1;
  }

  King::~King() {
    for (int i = 0; i < 8; ++i) {
      delete[] this->leaper_offsets.offsets[0][i];
    }
    delete[] this->leaper_offsets.offsets[0];
    delete[] this->leaper_offsets.offsets;
  }

  PogoPiece::PogoPiece(int id, int color) : Piece(id, color) {
    // reverse is map<number, number>
    this->leaper = 1;
    this->rotational_move_type = 2; // REVERSE ROTATE

    this->leaper_offsets.offsets = new int**[4];
    for (int i = 0; i < 4; ++i) {
      this->leaper_offsets.offsets[i] = new int*[2];
    }

    /* POGO PIECE MOVES */
    this->leaper_offsets.offsets[0][0][0] = 0;
    this->leaper_offsets.offsets[0][0][1] = 2;

    this->leaper_offsets.offsets[1][0][0] = 0;
    this->leaper_offsets.offsets[1][0][1] = -1;

    this->leaper_offsets.offsets[2][0][0] = 0;
    this->leaper_offsets.offsets[2][0][1] = -2;

    this->leaper_offsets.offsets[3][0][0] = 0;
    this->leaper_offsets.offsets[3][0][1] = 1;
    
    this->leaper_offsets.size = 4;
  }

  PogoPiece::~PogoPiece() {
    for (int i = 0; i < 4; ++i) {
      delete[] this->leaper_offsets.offsets[i];
    }
    delete[] this->leaper_offsets.offsets;
  }

  int Piece::get_id() {
    return this->id;
  }

  int Piece::get_king() {
    return this->king;
  }
  
  int Piece::get_pawn() {
    return this->pawn;
  }

  int Piece::get_enpassant() {
    return this->enpassant;
  }

  int Piece::get_promote() {
    return this->promote;
  }

  int Piece::get_color() {
    return this->color;
  }

  int Piece::get_leaper() {
    return this->leaper;
  }

  int Piece::get_slider() {
    return this->slider;
  }

  int Piece::get_straight() {
    return this->straight;
  }

  int Piece::get_diagonal() {
    return this->diagonal;
  }

  int Piece::get_rotational_move_type() {
    return this->rotational_move_type;
  }

  LeaperOffsets Piece::get_leaper_offsets() {
    return this->leaper_offsets;
  }

  int * Piece::get_straight_constraints() {
    return this->straight_constraints;
  }

  int * Piece::get_diagonal_constraints() {
    return this->diagonal_constraints;
  }

  uint64_t * Piece::get_diagonal_piece_mask() {
    return this->diagonal_piece_mask;
  }

  uint64_t * Piece::get_straight_piece_mask() {
    return this->straight_piece_mask;
  }

  uint64_t *** Piece::get_leaper_piece_state() {
    return this->leaper_piece_state;
  }

  uint64_t ** Piece::get_sliding_diagonal_piece_state() {
    return this->sliding_diagonal_piece_state;
  }

  uint64_t ** Piece::get_sliding_straight_piece_state() {
    return this->sliding_straight_piece_state;
  }

  uint64_t ** Piece::get_pawn_piece_state() {
    return this->pawn_piece_state;
  }

  uint64_t Piece::get_sliding_piece_attacks(int pos, uint64_t occupancy,
    int * straight_constraints, int * diagonal_constraints) {
    uint64_t piece_state = 0ULL;

    if (get_straight()) {
      uint64_t straight_occupancy = occupancy & get_straight_piece_mask()[pos];
      straight_occupancy *= mask_state::straight_magic_numbers[pos]
        >> (64 - mask_state::straight_relevant_bits[pos]);

      uint64_t raw_straight_moves = get_sliding_straight_piece_state()[pos][straight_occupancy];
      piece_state |= generator::apply_constraints(raw_straight_moves, straight_constraints, pos, 1);
    }

    if (get_diagonal()) {
      uint64_t diagonal_occupancy = occupancy & get_diagonal_piece_mask()[pos];
      diagonal_occupancy *= mask_state::diagonal_magic_numbers[pos]
        >> (64 - mask_state::diagonal_relevant_bits[pos]);
      
      uint64_t raw_diagonal_moves = get_sliding_diagonal_piece_state()[pos][diagonal_occupancy];
      piece_state |= generator::apply_constraints(raw_diagonal_moves, diagonal_constraints, pos, 0);
    }

    return piece_state;
  }

  uint64_t Piece::mask_leaper_attacks(int pos, int ** offsets, int color) {
    uint64_t current_attacks = 0ULL;
    uint64_t current_bitboard = 0ULL;

    current_bitboard = generator::set_bit(current_bitboard, pos);
    int is_black = color == 1;

    for (int i = 0; offsets[i] != nullptr; i++) {
      int file_offset = offsets[i][0];
      int rank_offset = offsets[i][1];

      if (is_black) {
        file_offset *= -1;
        rank_offset *= -1;
      }

      int64_t shift = static_cast<int64_t>(rank_offset) * 8 - file_offset;
      uint64_t file_constraint = generator::get_file_constraint(file_offset);

      if (shift > 0) {
        if (current_bitboard >> shift & file_constraint) {
          uint64_t shifted_bitboard = current_bitboard >> shift;
          current_attacks |= shifted_bitboard & file_constraint;
        }
      } else {
        if (current_bitboard << -shift & file_constraint) {
          uint64_t shifted_bitboard = current_bitboard << -shift;
          current_attacks |= shifted_bitboard & file_constraint;
        }
      }
    }
    
    return current_attacks;
  }

  uint64_t Piece::mask_pawn_attacks(int color, int pos) {
    uint64_t current_attacks = 0ULL;
    uint64_t current_bitboard = 0ULL;

    current_bitboard = generator::set_bit(current_bitboard, pos);
    int pawn = get_pawn();

    if (pawn != 0 && color == 0) {
      if (current_bitboard >> 7 & NOT_A_FILE)
        current_attacks |= current_bitboard >> 7;
      if (current_bitboard >> 9 & NOT_H_FILE)
        current_attacks |= current_bitboard >> 9;
    } else if (pawn != 0 && color == 1) {
      if (current_bitboard << 7 & NOT_H_FILE)
        current_attacks |= current_bitboard << 7;
      if (current_bitboard << 9 & NOT_A_FILE)
        current_attacks |= current_bitboard << 9;
    }

    return current_attacks;
  }

  uint64_t *** Piece::init_leaper_attacks() {
    LeaperOffsets leaper_offsets = get_leaper_offsets();
    size_t first_dim = leaper_offsets.size;
    uint64_t ***piece_state = new uint64_t**[first_dim];
    
    for (size_t i = 0; i < first_dim; i++) {
        piece_state[i] = new uint64_t*[LEAPER_PIECE_SIZE];
        for (size_t j = 0; j < LEAPER_PIECE_SIZE; j++) {
            piece_state[i][j] = new uint64_t[LEAPER_PIECE_STATE_INNER_SIZE]();
        }
    }

    for (size_t i = 0; i < first_dim; i++) {
        for (int square = 0; square < 64; square++) {
          piece_state[i][0][square] = mask_leaper_attacks(square, reinterpret_cast<int**>(leaper_offsets.offsets[i]), 0);
          piece_state[i][1][square] = mask_leaper_attacks(square, reinterpret_cast<int**>(leaper_offsets.offsets[i]), 1);
        }
    }

    return piece_state;
  }

  uint64_t ** Piece::init_pawn_attacks() {
    uint64_t ** piece_state = get_pawn_piece_state();

    for (int square = 0; square < 64; square++) {
      piece_state[0][square] = mask_pawn_attacks(0, square);
      piece_state[1][square] = mask_pawn_attacks(1, square);
    }

    return piece_state;
  }
}